wasm/text/func_parser.go

```diff
@@ -0,0 +1,131 @@
+package text
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/tetratelabs/wazero/wasm"
+	"github.com/tetratelabs/wazero/wasm/leb128"
+)
+
+// funcParser parses any instructions and dispatches to onBodyEnd.
+//
+// Ex.  `(module (func (nop)))`
+//       starts here --^    ^
+// onBodyEnd resumes here --+
+//
+// Note: funcParser is reusable. The caller resets when reaching the appropriate tokenRParen via beginBody.
+type funcParser struct {
+	// m is used as a function pointer to moduleParser.tokenParser. This updates based on state changes.
+	m *moduleParser
+
+	// onBodyEnd is called when complete parsing the body. Unless testing, this should be moduleParser.parseFuncEnd
+	onBodyEnd tokenParser
+
+	currentInstruction wasm.Opcode
+	// currentParameters are the current parameters to currentInstruction in WebAssembly 1.0 (MVP) binary format
+	currentParameters []byte
+
+	// currentCode is the current function body encoded in WebAssembly 1.0 (MVP) binary format
+	currentCode []byte
+}
+
+var end = []byte{wasm.OpcodeEnd}
+
+func (p *funcParser) getCode() []byte {
+	if p.currentCode == nil {
+		return end
+	}
+	return append(p.currentCode, wasm.OpcodeEnd)
```

Comment:
better add comment about the reason why we always end with end optcode: https://www.w3.org/TR/wasm-core-1/#expressions%E2%91%A0